Bump level

mjbain151 opened the pull request:
I added the ability to select the SemVer release level by adding the --level option.  One of my goals during implementation was to perverse the existing interface so that the code is backwards compatible with existing code.  The level option defaults to patch release so the program should behave exactly as it did before with no level option.  Please review and provide feedback.  Thanks

logicminds commented:
Please add change to the Changelog.md file

mjbain151 commented:
Updated the changelog.

logicminds commented on lib/release_manager/release.rb:
```diff
@@ -6,25 +6,30 @@ class Release
+  def level
+     options[:level] || 'patch'
+  end
+
```
I would put this in the cli code

`options[:level] ||= 'patch'`

logicminds commented on lib/release_manager/cli/release_mod_cli.rb:
```diff
@@ -38,6 +38,9 @@ def self.run
+        end
         opts.on('-m', '--module-path [MODULEPATH]', "The path to the module, defaults to #{Dir.getwd}") do |c|
           options[:path] = c
         end
```
Put this here: `options[:level] ||= 'patch'` after the opts.on

logicminds replied:
In order to store the option you have use the following syntax.  `opts.on('-l', '--level [LEVEL]', 'blah blah'`

Also in your description please separate the levels with commas `( major,minor,patch)`

Additionally, you should also supply some validation  %w(major minor patch).include?(c)

mjbain151 commented:
Agreed.  I will make the changes.

On Tue, Sep 26, 2017 at 2:23 PM, Corey Osman <notifications@github.com>
wrote:

> *@logicminds* commented on this pull request.
> ------------------------------
>
> In lib/release_manager/cli/release_mod_cli.rb
> <https://github.com/nwops/release_manager/pull/4#discussion_r141141791>:
>
> > @@ -38,6 +38,9 @@ def self.run
>          opts.on('-a', '--auto', 'Run this script without interaction') do |c|
>            options[:auto] = c
>          end
> +        opts.on('-l', '--level', 'Semantic versioning level to bump (major.minor.patch), defaults to patch') do |c|
> +          options[:level] = c
> +        end
>          opts.on('-m', '--module-path [MODULEPATH]', "The path to the module, defaults to #{Dir.getwd}") do |c|
>            options[:path] = c
>          end
>
> Put this here: options[:level] ||= 'patch' after the opts.on
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#pullrequestreview-65313584>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbEq-gLWuIambm9KMkFJ7rqjEroSuks5smUEIgaJpZM4PjA0d>
> .
>

logicminds commented on lib/release_manager/release.rb:
```diff
@@ -41,7 +46,9 @@ def bump
     end
     raise TagExists.new("Tag v#{version} already exists") if puppet_module.tag_exists?("v#{next_version}", options[:remote])
-    version = puppet_module.bump_patch_version unless options[:bump]
+    if puppet_module.respond_to?("bump_#{level}_version")
```
in order to check if a method exists you must convert it to a symbol.  This is the case for respond_to? and send()

`"bump_#{options[:level]}_version".to_sym`

mjbain151 replied:
Not sure I understand.  If level is a property containing the string value  "patch".  Then I would expect "bump_#{level}_version to return the string 'bump_patch_version'

GrammatonKlaric replied:
@mjbain151 See this doc [respond_to](https://ruby-doc.org/core-2.4.2/Object.html#method-i-respond_to-3F). Also check out [this nice article on ruby symbols](http://www.troubleshooters.com/codecorn/ruby/symbols.htm#_How_are_symbols_implemented)

logicminds commented:
https://github.com/nwops/release_manager/tree/instructions#develpment-environment-setup

mjbain151 commented:
I don't have docker installed.  So I will get that setup,  I don't know how
long that will take but will respond when I am done.

On Tue, Sep 26, 2017 at 4:19 PM, Corey Osman <notifications@github.com>
wrote:

> https://github.com/nwops/release_manager/tree/instructions#develpment-
> environment-setup
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-332322894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbLyUC6YRN6O4VTPUHEzFpIwAhzUsks5smVxCgaJpZM4PjA0d>
> .
>

mjbain151 commented:
So I got docker installed,  but executing docker-compose up  fails.

[b1mjbcd@b1lxdev02 release_manager]$ sudo docker version
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:06 2017
 OS/Arch:      linux/amd64
Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:00:25 2017
 OS/Arch:      linux/amd64
 Experimental: false
[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
-bash: docker-compose: command not found

On Tue, Sep 26, 2017 at 4:58 PM, Martin Bain <mjbain151@gmail.com> wrote:

> I don't have docker installed.  So I will get that setup,  I don't know
> how long that will take but will respond when I am done.
>
> On Tue, Sep 26, 2017 at 4:19 PM, Corey Osman <notifications@github.com>
> wrote:
>
>> https://github.com/nwops/release_manager/tree/instructions#
>> develpment-environment-setup
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nwops/release_manager/pull/4#issuecomment-332322894>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/Ac3NbLyUC6YRN6O4VTPUHEzFpIwAhzUsks5smVxCgaJpZM4PjA0d>
>> .
>>
>
>

logicminds commented:
Did you download docker for windows?

mjbain151 commented:
No my development server is a Linux server running Centos.7

On Tue, Sep 26, 2017 at 8:07 PM, Corey Osman <notifications@github.com>
wrote:

> Did you download docker for windows?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-332370693>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbNf4hse2v8HDDjxye27i2o3XkALDks5smZHdgaJpZM4PjA0d>
> .
>

logicminds commented:
You will need to install docker-compose as well.

mjbain151 commented:
I installed docker-compose as you suggested.  But is running into an issue
that others has encountered,  see below.  I status the docker deamon and it
is running.  So I am going to try different things to see if I can get to
work.

[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket -
is it running?
If it's at a non-standard location, specify the URL with the DOCKER_HOST
environment variable.

On Tue, Sep 26, 2017 at 10:17 PM, Corey Osman <notifications@github.com>
wrote:

> You will need to install docker-compose as well.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-332389125>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbKlTwcaHtt4aXGvnLugveFGav3NSks5smbAigaJpZM4PjA0d>
> .
>

mjbain151 commented:
So I resolve the issue with docker-compose not connecting to the docker
daemon.  That got me further along, but now I have a new issue to resolve.

[b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
Creating network "releasemanager_default" with the default driver
Pulling runner (
registry.gitlab.com/gitlab-org/gitlab-ci-multi-runner:ubuntu-v9.3.0)...
ERROR: Get
https://registry.gitlab.com/v2/gitlab-org/gitlab-ci-multi-runner/manifests/ubuntu-v9.3.0:
denied: access forbidden

On Wed, Sep 27, 2017 at 9:05 AM, Martin Bain <mjbain151@gmail.com> wrote:

> I installed docker-compose as you suggested.  But is running into an issue
> that others has encountered,  see below.  I status the docker deamon and it
> is running.  So I am going to try different things to see if I can get to
> work.
>
> [b1mjbcd@b1lxdev02 release_manager]$ docker-compose up
> ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket
> - is it running?
> If it's at a non-standard location, specify the URL with the DOCKER_HOST
> environment variable.
>
> On Tue, Sep 26, 2017 at 10:17 PM, Corey Osman <notifications@github.com>
> wrote:
>
>> You will need to install docker-compose as well.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nwops/release_manager/pull/4#issuecomment-332389125>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/Ac3NbKlTwcaHtt4aXGvnLugveFGav3NSks5smbAigaJpZM4PjA0d>
>> .
>>
>
>

logicminds commented:
I failed to mention that you needed some updates.  I pushed my changes to master.  Please rebase and try docker-compose up again.  There were some changes that you did not have.

mjbain151 commented:
So I did what you said and WOW! the magic started.  But alas it came to
halt when runner_1 started to throw warnings, see attached.

[image: Inline image 2]

On Wed, Sep 27, 2017 at 6:44 PM, Corey Osman <notifications@github.com>
wrote:

> I failed to mention that you needed some updates. I pushed my changes to
> master. Please rebase and try docker-compose up again. There were some
> changes that you did not have.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-332676787>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbBLjKHlcV8VoKnO8pjjAjpjhaFxVks5sms_igaJpZM4PjA0d>
> .
>

mjbain151 commented:
A short while later the errors went away and I was able to log into
GitLab-CE in the docker container.  So I have some exploring to do...
Ultimately I will need to get back to testing Release-manager.

[image: Inline image 1]

On Thu, Sep 28, 2017 at 6:11 AM, Martin Bain <mjbain151@gmail.com> wrote:

> So I did what you said and WOW! the magic started.  But alas it came to
> halt when runner_1 started to throw warnings, see attached.
>
> [image: Inline image 2]
>
> On Wed, Sep 27, 2017 at 6:44 PM, Corey Osman <notifications@github.com>
> wrote:
>
>> I failed to mention that you needed some updates. I pushed my changes to
>> master. Please rebase and try docker-compose up again. There were some
>> changes that you did not have.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nwops/release_manager/pull/4#issuecomment-332676787>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/Ac3NbBLjKHlcV8VoKnO8pjjAjpjhaFxVks5sms_igaJpZM4PjA0d>
>> .
>>
>
>

mjbain151 commented:
Added input validation for the --level parameter and squashed my commits.

On Thu, Sep 28, 2017 at 7:58 PM, Corey Osman <notifications@github.com>
wrote:

> Please squash your commits and rebase again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-332994982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbA2q4XDSm8XVMiUc9qSAbeiSyCH_ks5snDKVgaJpZM4PjA0d>
> .
>

mjbain151 commented:
Made some progress with docker.  But  I am getting there error below.
appuser@de96629fe060:/app$ bash app_startup_script.sh
Using rake 12.1.0
Using ast 2.3.0
Using bundler 1.16.0.pre.2
Using coderay 1.1.2
Using diff-lcs 1.3
Using multi_xml 0.6.0
Using httparty 0.15.6
Errno::EACCES: Permission denied @ rb_sysopen -
/usr/local/bundle/bin/httparty
An error occurred while installing httparty (0.15.6), and Bundler cannot
continue.
Make sure that `gem install httparty -v '0.15.6'` succeeds before bundling.
In Gemfile:
  release_manager was resolved to 0.5.3, which depends on
    gitlab was resolved to 4.2.0, which depends on
      httparty
Identity added: /home/appuser/.ssh/id_rsa (rsa w/o comment)


On Mon, Oct 2, 2017 at 4:13 PM, Martin Bain <mjbain151@gmail.com> wrote:

> Added input validation for the --level parameter and squashed my commits.
>
> On Thu, Sep 28, 2017 at 7:58 PM, Corey Osman <notifications@github.com>
> wrote:
>
>> Please squash your commits and rebase again.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nwops/release_manager/pull/4#issuecomment-332994982>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/Ac3NbA2q4XDSm8XVMiUc9qSAbeiSyCH_ks5snDKVgaJpZM4PjA0d>
>> .
>>
>
>

logicminds commented:
Yea I was getting that error too.  I thought I resolved it.  Also your commits message needs to be summarized.  Please follow this as a guide: https://docs.puppet.com/forge/contributing.html#making-changes

mjbain151 commented:
Lots of interesting tips in the guide.  But I don't think I found anything
on how to summarize your commit messages.  In fact I saw interesting
articles on "Commit early and often" and "On the sausage making",   Were
you referring to squashing my commits?

On Mon, Oct 2, 2017 at 5:05 PM, Corey Osman <notifications@github.com>
wrote:

> Yea I was getting that error too. I thought I resolved it. Also your
> commits message needs to be summarized. Please follow this as a guide:
> https://docs.puppet.com/forge/contributing.html#making-changes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nwops/release_manager/pull/4#issuecomment-333664985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac3NbJL3k_ZCoesU9z2IHfgs4KPkQaTUks5soVAfgaJpZM4PjA0d>
> .
>

logicminds commented:
Yes please summary your 15 or so commit messages into a few sentences.  What does this PR provide and why it didn't work before.

mjbain151 commented:
Amended the commits into a single commit summary.  Thanks.

On Tue, Oct 3, 2017 at 1:51 PM, Martin Bain <mjbain151@gmail.com> wrote:

> Lots of interesting tips in the guide.  But I don't think I found anything
> on how to summarize your commit messages.  In fact I saw interesting
> articles on "Commit early and often" and "On the sausage making",   Were
> you referring to squashing my commits?
>
> On Mon, Oct 2, 2017 at 5:05 PM, Corey Osman <notifications@github.com>
> wrote:
>
>> Yea I was getting that error too. I thought I resolved it. Also your
>> commits message needs to be summarized. Please follow this as a guide:
>> https://docs.puppet.com/forge/contributing.html#making-changes
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nwops/release_manager/pull/4#issuecomment-333664985>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/Ac3NbJL3k_ZCoesU9z2IHfgs4KPkQaTUks5soVAfgaJpZM4PjA0d>
>> .
>>
>
>
